acceptance/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,79 @@
+package upgrade_test
+
+import (
+	"net/http"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/epinio/epinio/acceptance/helpers/catalog"
+	"github.com/epinio/epinio/acceptance/helpers/proc"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Epinio upgrade with running app", func() {
+	var (
+		namespace string
+		appName   string
+		domain    string
+	)
+
+	BeforeEach(func() {
+		namespace = catalog.NewNamespaceName()
+		env.SetupAndTargetNamespace(namespace)
+		appName = catalog.NewAppName()
+		domain = os.Getenv("EPINIO_SYSTEM_DOMAIN")
+	})
+
+	AfterEach(func() {
+		env.DeleteApp(appName)
+		env.DeleteNamespace(namespace)
+	})
+
+	It("can upgrade epinio", func() {
+		// Deploy a simple application before upgrading Epinio
+		out := env.MakeGolangApp(appName, 1, true)
+		routeRegexp := regexp.MustCompile(`https:\/\/.*omg.howdoi.website`)
+		route := string(routeRegexp.Find([]byte(out)))
+
+		// Check that the app is reachable
+		Eventually(func() int {
+			resp, err := env.Curl("GET", route, strings.NewReader(""))
+			Expect(err).ToNot(HaveOccurred())
+			return resp.StatusCode
+		}, 30*time.Second, 1*time.Second).Should(Equal(http.StatusOK))
+
+		// Build image with latest epinio-server binary
+		_, err := proc.Run("../..", false, "docker", "build", "-t", "epinio/epinio-server", "-f", "images/Dockerfile", ".")
```

Comment:
strangely, this has happened already [when we called prepare_environment_k3d](https://github.com/epinio/epinio/pull/1607/files#diff-4ead13f0ea5e43870e3dc8b9fc7a6d1220ff0a5bb8e00b2db8ae1f5ae73b116cR71) because that target [also calls build-images](https://github.com/epinio/epinio/blob/e6c2b297d211ad5854f5c380128991e9ea6d2f3f/Makefile#L187). But we should probably not rely on that because we don't know here if that target was called and what it looks like. It's just that we already rely on the "prepare" make target to build the binary (the dockerfile simply copies that).

The setup of this test is happening in 3 different places: this script, the Makefile and the pipeline. Otoh, this script is completely optional, we just use Go to have nice consistent output. I'm not sure how to better structure it. We can keep it as it is until we have a better idea.

Reply:
yes, I saw that the `prepare_environment_k3d` calls the build-images target. I agree that could be improved later, I only wanted to have a running upgrade test and especially have a correct process.